### PR TITLE
feature(wysiwyg): Support for syntax highlighting

### LINKED ISSUE #43: Make it easier to post code snippets
Good grief this has been broken way too long. Personally I'd like typing 3 backticks + `php` + enter to be replaced by an HTML code block that is highlighted as PHP.

But I'd also be willing to settle for a simple code-block button :)

### PR DESCRIPTION
**Pull request by ewinslow**
Fixes #43 

TODOs:
 * [x] Merge Elgg/Elgg#8804
 * [x] Seems to be breaking discussion-reply editing (Resolution: Disable ajax reply editing)
 * [x] Dragging code blocks throws an error (Resolution: Disable "blockimagepaste")
 * [x] Edit HTML button doesn't work

### CODE REVIEW

**Comment by ewinslow** [comment]
Before:

code-snippets-before

After:

code-snippets-after

**Comment by ewinslow** [comment]
Would rather not submit until https://github.com/Elgg/Elgg/pull/8805 and https://github.com/Elgg/Elgg/pull/8804 are submitted, so we don't have to point at a custom branch of elgg/elgg.

File: start.php [hunk 1]
```diff
@@ -7,21 +7,39 @@ function init() {
...
 		http_response_code(301);
 		forward('http://learn.elgg.org/en/latest/intro/license.html');
 	});
+
+
+	// Support for syntax-highlighting
+	elgg_extend_view('elgg.css', 'pre.css');
+
+	elgg_register_plugin_hook_handler('spec', 'htmlawed', function($hook, $type, $result, $params) {
+		return "$result;code=class(oneof=language-php|language-css|language-javascript)";
```
Comment (juho-jaakkola) [inline start.php]: So this allows using elements like this?
```html
<code class="language-php">$foo = bar;</code>
```
And these are created by a ckeditor plugin?

Maybe worth adding some more comments on what exactly is being done and why.

File: start.php [hunk 2]
```diff
@@ -7,21 +7,39 @@ function init() {
...
+	});
+
+	elgg_register_css('highlight', elgg_get_simplecache_url('ckeditor/plugins/codesnippet/lib/highlight/styles/default.css'));
+	elgg_load_css('highlight');
+
+	elgg_define_js('highlight', [
+		'src' => elgg_get_simplecache_url('ckeditor/plugins/codesnippet/lib/highlight/highlight.pack.js'),
```
Comment (juho-jaakkola) [inline start.php]: CKEditor has this plugin by default?

Comment (ewinslow) [inline start.php]: The latest one pulled down from composer does. Should probably update the one in elgg/elgg. Or extract it to a separate plugin so we don't have to check in ckeditor to elgg/elgg in the first place.

File: start.php (hunk 1 above)
Comment (ewinslow) [inline start.php]: Done

**Comment by mrclay** [comment]
:+1: :+1: :+1:

**Comment by ewinslow** [comment]
So I'm thinking we just submit this without waiting for htmlawed-spec PR to land (since it has to wait until 2.1). The ability to specify language would be nice, but it's not a deal-breaker, and this is already much better than it was before. Highlight-js is pretty good about highlighting codeblocks that don't have a language specified, too.

**Comment by juho-jaakkola** [comment]
IIRC core allows registering for non-existing hooks. So we can just merge this, and the spec hook will start working once community gets upgraded to 2.1.

**Comment by ewinslow** [comment]
Yep. Need to figure out discussion replies, but then we should be good to go.

File: composer.json [hunk 3]
```diff
@@ -50,6 +52,10 @@
         },
         {
             "type": "vcs",
+            "url": "https://github.com/ewinslow/elgg"
```
Comment (juho-jaakkola) [inline composer.json]: Not needed anymore

**Comment by ewinslow** [comment]
Couple more comments / improvements: 

 * I made the editor inline instead of in an iframe. This way the toolbar doesn't take up space of you're not using the edit area.

 * I'm getting errors when I try to drag code blocks around, and this feature is basically unusable on mobile.

* editing code blocks requires opening another dialog, and that code is not highlighted.

* On mobile you have to press the return/enter button to edit a code block. Double tap doesn't seem to work.

* there's an odd focus outline bug when editing with the current inline editor. Or has something to do with the fact that I added the elgg-output class which I needed to do in order to get list styles to appear correctly.

**Comment by ewinslow** [comment]
In order to fix the remaining bugs, I had to:

 * Switch to non-ajax for discussion reply editing
 * Disable the "blockimagepaste" plugin

How do we feel about those sacrifices?

I'm hopeful they represent rare enough operations that it's not too big a sacrifice... Nothing else is ajax edited and we live. I feel like the inability to display code reasonable is a much more significant shortcoming ATM.

I let you all weigh in before I pull any triggers or anything.

**Comment by ewinslow** [comment]
Ugh... this has been quite frustrating. The Edit HTML button doesn't work now... this thing feels so fragile...

File: composer.json (hunk 3 above)
Comment (ewinslow) [inline composer.json]: Done

File: start.php [hunk 4]
```diff
@@ -25,23 +25,45 @@
...
+	// These are generated by the ckeditor "codesnippet" plugin
+	// NB: Won't have any effect until https://github.com/Elgg/Elgg/pull/8805 lands.
+	elgg_register_plugin_hook_handler('spec', 'htmlawed', function($hook, $type, $result, $params) {
+		return "$result;code=class(oneof=language-php|language-css|language-javascript)";
+	});
+
+	// Load the JS and CSS to get syntax highlighting for rendered HTML
+	elgg_define_js('hljs', [
+		'src' => elgg_get_simplecache_url('ckeditor/plugins/codesnippet/lib/highlight/highlight.pack.js'),
+		'deps' => ['css!ckeditor/plugins/codesnippet/lib/highlight/styles/default.css'],
```
Comment (mrclay) [inline start.php]: Does the CSS path always resolve from the install root?

**Comment by ewinslow** [comment]
Implemented my own little plugin to just apply the pre/code tags inline instead of opening up in a new dialog. I felt like that was just an overly complicated experience.

Went back to iframed editor, even though I do despise the "heavy" feel of the current editor style. It's complicating/breaking things too much to try to switch to inline.

Here's the view from the editor. I rearranged the toolbar while I was at it to be able to put the code-block button in a logical place.

code-block-editor

Editing code on mobile is substantially improved.

**Comment by ewinslow** [comment]
Also I removed the bottom bar, (even though it's shown in the screenshot above, I tweaked the CSS after the fact).

**Comment by ewinslow** [comment]
Reverted to the core version of CKEditor since we don't need the codesnippets plugin. Imported our own version of highlightjs for syntax highlighting since we can't use the one provided by codesnippets plugin.

There are a few quirks left that I need to figure out. E.g. if you change an empty paragraph to a code block, when you start typing, the content doesn't make it into the `<code></code>` tags.

File: start.php (hunk 4 above)
Comment (ewinslow) [inline start.php]: From the simplecache url root, yes.

**Comment by mrclay** [comment]
:+1:

**Comment by ewinslow** [comment]
Looks good! Excited to be able to share code snippets that aren't horrible...

**Pull request merged**